piedpiper_gman/gman.py

```diff
@@ -10,108 +12,287 @@
+
+        elif context == 'add_event':
+            disallowed = ('caller', 'thread_id', 'timestamp', 'project', 'run_id')
+
+            self._event = TaskEventSchema().load(self.raw_data, partial=('timestamp',))
+
+            try:
+                self._event.data.timestamp = datetime.datetime.now()
+            except AttributeError:
+                self._event.data['timestamp'] = datetime.datetime.now()
+
+            if self._event.errors:
+                self.errors.extend(self._event.errors)
+
+            if self.raw_data.get('status', '') in ('started', 'received'):
+                self.errors.add('status', 'Updates may not be value \'started\'')
+
+        for key in disallowed:
+            if key in self.raw_data:
+                self.errors.add(key, f'May not be specified for {context}')
+
+        if len(self.errors.errors):
+            raise MarshalError(self.errors)
+
+    @property
+    def task(self):
+        return self._task
+
+    @property
+    def event(self):
+        return self._event
+
+
+class GMan(Resource):
+
+    def get_task_states(self, events):
+        '''Parses task events to identify which state the task is in.'''
+
+        running = ('started', 'received')
+        completed = ('completed',)
+        failed = ('failed',)
+
+        states = {'running': [],
+                  'completed': [],
+                  'failed': []}
+
+        for event in events:
+            if event.status in running:
+                if event.task not in states['running']:
+                    states['running'].append(event.task)
+            elif event.status in completed:
+                if event.task not in states['completed']:
+                    states['completed'].append(event.task)
+                if event.task in states['running']:
+                    states['running'].remove(event.task)
+            elif event.status in failed:
+                if event.task not in states['failed']:
+                    states['failed'].append(event.task)
+                if event.task in states['running']:
+                    states['running'].remove(event.task)
+
+        return states
+
+    def get_event_thread(self, thread_id):
+        return [x for x in
+                TaskEvent.select()
+                         .join(Task)
+                         .where(Task.thread_id == thread_id)
+                         .order_by(TaskEvent.timestamp)]
+
+    def get_task_events(self, task_id):
+        return [x for x in
+                TaskEvent.select()
+                         .join(Task)
+                         .where(Task.task_id == task_id)
+                         .order_by(TaskEvent.timestamp)]
+
+    def head(self, thread_id=None, events=None, task_id=None):
+
+        if thread_id:
+            task_events = self.get_event_thread(thread_id)
+        elif task_id:
+            task_events = self.get_task_events(task_id)
+
+        if len(task_events):
```

Comment:
Should we check if task_events exists before checking its length? Or maybe set it to empty first. Currently gman will return a 500 using:
curl --head localhost/gman

Reply:
yes, yes we should

Reply:
fixed although it's not updating showing that I did... IDK